doc/man/syslog-ng-ctl.1.in
'\" t
.\"     Title: syslog-ng-ctl
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 07/05/2016
.\"    Manual: The syslog-ng control tool manual page
.\"    Source: 5 F6
.\"  Language: English
.\"
.TH "SYSLOG\-NG\-CTL" "1" "07/05/2016" "5 F6" "The syslog-ng control tool man"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
syslog-ng-ctl \- Display message statistics and enable verbose, debug and trace modes in
.SH "SYNOPSIS"
.HP \w'\fBsyslog\-ng\-ctl\fR\ 'u
\fBsyslog\-ng\-ctl\fR [command] [options]
.SH "DESCRIPTION"
.PP
NOTE: The syslog\-ng\-ctl application is distributed with the system logging application, and is usually part of the syslog\-ng package\&. The latest version of the syslog\-ng application is available at
\m[blue]\fBthe official syslog\-ng website\fR\m[]\&.
.PP
This manual page is only an abstract; for the complete documentation of syslog\-ng, see
\m[blue]\fBThe syslog\-ng Administrator Guide\fR\m[]\&.
.PP
The
\fBsyslog\-ng\-ctl\fR
application is a utility that can be used to:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
enable/disable various syslog\-ng messages for troubleshooting;
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
display statistics about the processed messages;
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
reload the configuration of ;
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
stop \&.
.RE
.SH "ENABLING TROUBLESHOOTING MESSAGES"
.HP \w'\fBcommand\fR\ 'u
\fBcommand\fR [options]
.PP
Use the
\fBsyslog\-ng\-ctl <command> \-\-set=on\fR
command to display verbose, trace, or debug messages\&. If you are trying to solve configuration problems, the verbose (and occasionally trace) messages are usually sufficient; debug messages are needed mostly for finding software errors\&. After solving the problem, do not forget to turn these messages off using the
\fBsyslog\-ng\-ctl <command> \-\-set=off\fR\&. Note that enabling debug messages does not enable verbose and trace messages\&.
.PP
Use
\fBsyslog\-ng\-ctl <command>\fR
without any parameters to display whether the particular type of messages are enabled or not\&.
.PP
If you need to use a non\-standard control socket to access syslog\-ng, use the
\fBsyslog\-ng\-ctl <command> \-\-set=on \-\-control=<socket>\fR
command to specify the socket to use\&.
.PP
\fBverbose\fR
.RS 4
Print verbose messages\&. If syslog\-ng was started with the
\fI\-\-stderr\fR
or
\fB\-e\fR
option, the messages will be sent to stderr\&. If not specified, syslog\-ng will log such messages to its internal source\&.
.RE
.PP
\fBtrace\fR
.RS 4
Print trace messages of how messages are processed\&. If syslog\-ng was started with the
\fI\-\-stderr\fR
or
\fB\-e\fR
option, the messages will be sent to stderr\&. If not specified, syslog\-ng will log such messages to its internal source\&.
.RE
.PP
\fBdebug\fR
.RS 4
Print debug messages\&. If syslog\-ng was started with the
\fI\-\-stderr\fR
or
\fB\-e\fR
option, the messages will be sent to stderr\&. If not specified, syslog\-ng will log such messages to its internal source\&.
.RE
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
syslog\-ng\-ctl verbose \-\-set=on
.fi
.if n \{\
.RE
.\}
.SH "THE STATS COMMAND"
.HP \w'\fBstats\fR\ 'u
\fBstats\fR [options]
.PP
Use the
\fBstats\fR
command to display statistics about the processed messages\&. For details about the displayed statistics, see
\m[blue]\fBThe syslog\-ng Administrator Guide\fR\m[]???\&. The
\fBstats\fR
command has the following options:
.PP
\fB\-\-control=<socket>\fR or \fB\-c\fR
.RS 4
Specify the socket to use to access syslog\-ng\&. Only needed when using a non\-standard socket\&.
.RE
.PP
\fB\-\-reset=<socket>\fR or \fB\-r\fR
.RS 4
Reset all statistics to zero, except for the
\fBstored\fR
counters\&. (The
\fBstored\fR
counters show the number of messages stored in the message queue of the destination driver, waiting to be sent to the destination\&.)
.RE
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
syslog\-ng\-ctl stats
.fi
.if n \{\
.RE
.\}
.PP
An example output:
.sp
.if n \{\
.RS 4
.\}
.nf
src\&.internal;s_all#0;;a;processed;6445
src\&.internal;s_all#0;;a;stamp;1268989330
destination;df_auth;;a;processed;404
destination;df_news_dot_notice;;a;processed;0
destination;df_news_dot_err;;a;processed;0
destination;d_ssb;;a;processed;7128
destination;df_uucp;;a;processed;0
source;s_all;;a;processed;7128
destination;df_mail;;a;processed;0
destination;df_user;;a;processed;1
destination;df_daemon;;a;processed;1
destination;df_debug;;a;processed;15
destination;df_messages;;a;processed;54
destination;dp_xconsole;;a;processed;671
dst\&.tcp;d_network#0;10\&.50\&.0\&.111:514;a;dropped;5080
dst\&.tcp;d_network#0;10\&.50\&.0\&.111:514;a;processed;7128
dst\&.tcp;d_network#0;10\&.50\&.0\&.111:514;a;stored;2048
destination;df_syslog;;a;processed;6724
destination;df_facility_dot_warn;;a;processed;0
destination;df_news_dot_crit;;a;processed;0
destination;df_lpr;;a;processed;0
destination;du_all;;a;processed;0
destination;df_facility_dot_info;;a;processed;0
center;;received;a;processed;0
destination;df_kern;;a;processed;70
center;;queued;a;processed;0
destination;df_facility_dot_err;;a;processed;0
.fi
.if n \{\
.RE
.\}
.SH "DISPLAYING LICENSE-RELATED INFORMATION"
.HP \w'\fBsyslog\-ng\-ctl\ show\-license\-info\fR\ 'u
\fBsyslog\-ng\-ctl show\-license\-info\fR [options]
.PP
The application uses a license in server mode to determine the maximum number of hosts that are allowed to connect\&. Use the
\fBsyslog\-ng\-ctl show\-license\-info\fR
command to display license\-related information the number of hosts currently logging to your server\&. This helps you to plan your capacity, to check your license usage, and to detect client misconfiguration that can result in a license miscount anomaly\&. Note that in client or relay mode, does not require a license\&.
.PP
The
\fBsyslog\-ng\-ctl show\-license\-info\fR
command displays the following information\&. In case of an unlimited license, or in client or relay mode, only the license type is displayed:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
License Type: none, limited, unlimited
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Host Limit: the maximum number of hosts that are allowed to connect\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Currently Used Slots: the number of currently used host slots
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Usage: the percent of used host slots
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Licensed Clients: the list of hostnames that are stored in the license module
.RE
.PP
The
\fBsyslog\-ng\-ctl show\-license\-info\fR
command has the following options:
.PP
\fB\-\-json\fR or \fB\-J\fR
.RS 4
Print license\-related information in JSON format\&.
.RE
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
syslog\-ng\-ctl show\-license\-info
.fi
.if n \{\
.RE
.\}
.PP
An example output:
.sp
.if n \{\
.RS 4
.\}
.nf
License\-Type: limited
Host\-Limit: 10
Currently\-Used\-Slots: 7
Usage: 70%
Licensed\-Clients:
    192\&.168\&.0\&.1
    192\&.168\&.0\&.2
    192\&.168\&.0\&.3
    192\&.168\&.1\&.4
    192\&.168\&.1\&.5
.fi
.if n \{\
.RE
.\}
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
syslog\-ng\-ctl show\-license\-info \-\-json
.fi
.if n \{\
.RE
.\}
.PP
An example output:
.sp
.if n \{\
.RS 4
.\}
.nf
{
    "license_type": "limited",
    "host_limit": 10,
    "currently_used_slots": 7,
    "usage": "70%",
    "licensed_clients": [
        "xy\&.testdomain",
        "testhost",
        "192\&.168\&.0\&.3",
        "test_host",
        "192\&.168\&.1\&.5"
    ]
}
.fi
.if n \{\
.RE
.\}
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
syslog\-ng\-ctl show\-license\-info
.fi
.if n \{\
.RE
.\}
.sp
in case of an unlimited license
.PP
An example output:
.sp
.if n \{\
.RS 4
.\}
.nf
$ syslog\-ng\-ctl show\-license\-info
License\-Type: unlimited
.fi
.if n \{\
.RE
.\}
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
syslog\-ng\-ctl show\-license\-info
.fi
.if n \{\
.RE
.\}
.sp
if is in client or relay mode
.PP
An example output:
.sp
.if n \{\
.RS 4
.\}
.nf
$ syslog\-ng\-ctl show\-license\-info
License\-Type: none
.fi
.if n \{\
.RE
.\}
.SH "RELOADING THE CONFIGURATION"
.HP \w'\fBsyslog\-ng\-ctl\ reload\fR\ 'u
\fBsyslog\-ng\-ctl reload\fR [options]
.PP
Use the
\fBsyslog\-ng\-ctl reload\fR
command to reload the configuration file of without having to restart the application\&. The
\fBsyslog\-ng\-ctl reload\fR
works like a SIGHUP (\-1)\&.
.SH "STOPPING"
.HP \w'\fBsyslog\-ng\-ctl\ stop\fR\ 'u
\fBsyslog\-ng\-ctl stop\fR [options]
.PP
Use the
\fBsyslog\-ng\-ctl stop\fR
command to stop the application\&. The
\fBsyslog\-ng\-ctl stop\fR
works like a SIGHUP (\-15) on Linux and Unix systems\&.
.SH "FILES"
.PP
/opt/syslog\-ng/sbin/syslog\-ng\-ctl
.SH "SEE ALSO"
.PP
\m[blue]\fB\fBThe syslog\-ng Administrator Guide\fR\fR\m[]
.PP
\fBsyslog\-ng\&.conf\fR(5)
.PP
\fBsyslog\-ng\fR(8)
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
For the detailed documentation of see
\m[blue]\fB\fBThe 5 F6 Administrator Guide\fR\fR\m[]
.PP
If you experience any problems or need help with syslog\-ng, visit
\m[blue]\fBvisit the syslog\-ng wiki\fR\m[]
or the
\m[blue]\fB\fBsyslog\-ng mailing list\fR\fR\m[]\&.
.PP
For news and notifications about of syslog\-ng, visit the
\m[blue]\fB\fBsyslog\-ng Blog\fR\fR\m[]\&.
.sp .5v
.RE
.SH "AUTHOR"
.PP
This manual page was written by the BalaBit Documentation Team <documentation@balabit\&.com>\&.
.SH "COPYRIGHT"


syslog-ng-ctl/syslog-ng-ctl.c
/*
 * Copyright (c) 2002-2012 Balabit
 * Copyright (c) 1998-2012 Balázs Scheidler
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include "syslog-ng.h"
#include "gsocket.h"
#include "control_client.h"
#include "cfg.h"
#include "reloc.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#if HAVE_GETOPT_H
#include <getopt.h>
#endif

static gchar *control_name = PATH_CONTROL_SOCKET;
static ControlClient *control_client;

static gboolean
slng_send_cmd(const gchar * const cmd)
{
  if (!control_client_connect(control_client))
    {
      return FALSE;
    }

  if (control_client_send_command(control_client,cmd) < 0)
    {
      return FALSE;
    }

  return TRUE;
}

static GString *
slng_run_command(const gchar * const command)
{
  if (slng_send_cmd(command))
    return control_client_read_reply(control_client);

  return NULL;
}

static gchar *verbose_set = NULL;

static gint
slng_verbose(int argc, char *argv[], const gchar *mode)
{
  gint ret = 0;
  GString *rsp = NULL;
  gchar buff[256];

  if (!verbose_set)
    snprintf(buff, 255, "LOG %s\n", mode);
  else
    snprintf(buff, 255, "LOG %s %s\n", mode,
        strncasecmp(verbose_set, "on", 2) == 0 || verbose_set[0] == '1' ? "ON" : "OFF");

  g_strup(buff);

  rsp = slng_run_command(buff);
  if (rsp == NULL)
    return 1;

  if (!verbose_set)
    printf("%s\n", rsp->str);
  else
    ret = g_str_equal(rsp->str, "OK");

  g_string_free(rsp, TRUE);

  return ret;
}

static gboolean stats_options_reset_is_set = FALSE;

static GOptionEntry stats_options[] =
{
  { "reset", 'r', 0, G_OPTION_ARG_NONE, &stats_options_reset_is_set, "reset counters", NULL },
  { NULL,    0,   0, G_OPTION_ARG_NONE, NULL,                        NULL,             NULL }
};

static GOptionEntry verbose_options[] =
{
  { "set", 's', 0, G_OPTION_ARG_STRING, &verbose_set,
    "enable/disable messages", "<on|off|0|1>" },
  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
};


static const gchar *
_stats_command_builder()
{
  return stats_options_reset_is_set ? "RESET_STATS\n" : "STATS\n";
}

static gint
slng_stats(int argc, char *argv[], const gchar *mode)
{
  GString *rsp = slng_run_command(_stats_command_builder());

  if (rsp == NULL)
    return 1;

  printf("%s\n", rsp->str);

  g_string_free(rsp, TRUE);

  return 0;
}

static gboolean license_json = FALSE;

static GOptionEntry license_options[] =
{
  { "json", 'J', 0, G_OPTION_ARG_NONE, &license_json, "enable json output", NULL },
  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
};

static gint
slng_license(int argc, char *argv[], const gchar *mode)
{
  GString *rsp = NULL;
  gchar buff[256];

  if (license_json)
    snprintf(buff, 255, "LICENSE JSON\n");
  else
    snprintf(buff, 255, "LICENSE\n");

  if (!(slng_send_cmd(buff) && ((rsp = control_client_read_reply(control_client)) != NULL)))
    return 1;

  printf("%s\n", rsp->str);

  g_string_free(rsp, TRUE);

  return 0;
}

static gint
slng_stop(int argc, char *argv[], const gchar *mode)
{
  GString *rsp = slng_run_command("STOP\n");

  if (rsp == NULL)
    return 1;

  printf("%s\n", rsp->str);

  g_string_free(rsp, TRUE);

  return 0;
}

static gint
slng_reload(int argc, char *argv[], const gchar *mode)
{
  GString *rsp = slng_run_command("RELOAD\n");

  if (rsp == NULL)
    return 1;

  printf("%s\n", rsp->str);

  g_string_free(rsp, TRUE);

  return 0;
}

static GOptionEntry no_options[] =
{
  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
};

const gchar *
get_mode(int *argc, char **argv[])
{
  gint i;
  const gchar *mode;

  for (i = 1; i < (*argc); i++)
    {
      if ((*argv)[i][0] != '-')
        {
          mode = (*argv)[i];
          memmove(&(*argv)[i], &(*argv)[i+1], ((*argc) - i) * sizeof(gchar *));
          (*argc)--;
          return mode;
        }
    }
  return NULL;
}

static GOptionEntry slng_options[] =
{
  { "control", 'c', 0, G_OPTION_ARG_STRING, &control_name,
    "syslog-ng control socket", "<socket>" },
  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
};

static struct
{
  const gchar *mode;
  const GOptionEntry *options;
  const gchar *description;
  gint (*main)(gint argc, gchar *argv[], const gchar *mode);
} modes[] =
{
  { "stats", stats_options, "Query/reset syslog-ng statistics", slng_stats },
  { "verbose", verbose_options, "Enable/query verbose messages", slng_verbose },
  { "debug", verbose_options, "Enable/query debug messages", slng_verbose },
  { "trace", verbose_options, "Enable/query trace messages", slng_verbose },
  { "stop", no_options, "Stop syslog-ng process", slng_stop },
  { "reload", no_options, "Reload syslog-ng", slng_reload },
  { "show-license-info", license_options, "Show information about the license", slng_license },
  { NULL, NULL },
};

void
usage(const gchar *bin_name)
{
  gint mode;

  fprintf(stderr, "Syntax: %s <command> [options]\nPossible commands are:\n", bin_name);
  for (mode = 0; modes[mode].mode; mode++)
    {
      fprintf(stderr, "    %-20s %s\n", modes[mode].mode, modes[mode].description);
    }
  exit(1);
}

#include "reloc.c"

int
main(int argc, char *argv[])
{
  const gchar *mode_string;
  GOptionContext *ctx;
  gint mode;
  GError *error = NULL;
  int result;

  control_name = get_reloc_string(PATH_CONTROL_SOCKET);

  mode_string = get_mode(&argc, &argv);
  if (!mode_string)
    {
      usage(argv[0]);
    }

  ctx = NULL;
  for (mode = 0; modes[mode].mode; mode++)
    {
      if (strcmp(modes[mode].mode, mode_string) == 0)
        {
          ctx = g_option_context_new(mode_string);
          #if GLIB_CHECK_VERSION (2, 12, 0)
          g_option_context_set_summary(ctx, modes[mode].description);
          #endif
          g_option_context_add_main_entries(ctx, modes[mode].options, NULL);
          g_option_context_add_main_entries(ctx, slng_options, NULL);
          break;
        }
    }
  if (!ctx)
    {
      fprintf(stderr, "Unknown command\n");
      usage(argv[0]);
    }

  if (!g_option_context_parse(ctx, &argc, &argv, &error))
    {
      fprintf(stderr, "Error parsing command line arguments: %s\n", error ? error->message : "Invalid arguments");
      g_clear_error(&error);
      g_option_context_free(ctx);
      return 1;
    }
  g_option_context_free(ctx);

  control_client = control_client_new(control_name);

  result = modes[mode].main(argc, argv, modes[mode].mode);
  control_client_free(control_client);
  return result;
}
